### PR TITLE
vmselect: allow to override the default unique time series limit in v…

### PR DESCRIPTION
**Pull request by Haleygo**
…mstorage using command-line flags like `-search.maxUniqueTimeseries`, `-search.maxLabelsAPISeries`

This behavior is prohibited starting from v1.105.0, with https://github.com/VictoriaMetrics/VictoriaMetrics/commit/85f60237e25d324819f4fd23a63274660f5bd560. 

address https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7852

### CODE REVIEW

File: app/vmstorage/servers/vmselect.go [hunk 1]
```diff
@@ -253,16 +253,16 @@ func checkTimeRange(s *storage.Storage, tr storage.TimeRange) error {
...
-	if maxMetricsLimit <= 0 {
-		maxMetricsLimit = GetMaxUniqueTimeSeries()
+func getMaxMetrics(searchQueryLimit int) int {
+	if searchQueryLimit <= 0 {
+		// use vmstorage maxUniqueTimeseries if no limit is passed by the client.
+		return GetMaxUniqueTimeSeries()
 	}
-	if maxMetrics <= 0 || maxMetrics > maxMetricsLimit {
-		maxMetrics = maxMetricsLimit
+	// if `-search.maxUniqueTimeseries` is explicitly set in vmstorage, client cannot exceed the limit.
```
Comment (hagen1778) [inline app/vmstorage/servers/vmselect.go:261]:
```suggestion
	// searchQueryLimit  can exceed value explicitly set to `-search.maxUniqueTimeseries`
```

this func is unaware of client/server terms

File: app/vmstorage/servers/vmselect.go [hunk 2]
```diff
@@ -253,16 +253,16 @@ func checkTimeRange(s *storage.Storage, tr storage.TimeRange) error {
...
 }
 
-func getMaxMetrics(sq *storage.SearchQuery) int {
-	maxMetrics := sq.MaxMetrics
-	maxMetricsLimit := *maxUniqueTimeseries
-	if maxMetricsLimit <= 0 {
-		maxMetricsLimit = GetMaxUniqueTimeSeries()
+func getMaxMetrics(searchQueryLimit int) int {
+	if searchQueryLimit <= 0 {
+		// use vmstorage maxUniqueTimeseries if no limit is passed by the client.
```
Comment (hagen1778) [inline app/vmstorage/servers/vmselect.go:258]:
```suggestion
		// use auto calculated maxUniqueTimeseries limit
```

File: app/vmselect/prometheus/prometheus.go [hunk 3]
```diff
@@ -54,7 +54,7 @@ var (
 	selectNodes = flagutil.NewArrayString("selectNode", "Comma-separated addresses of vmselect nodes; usage: -selectNode=vmselect-host1,...,vmselect-hostN")
 
 	maxUniqueTimeseries = flag.Int("search.maxUniqueTimeseries", 0, "The maximum number of unique time series, which can be selected during /api/v1/query and /api/v1/query_range queries. This option allows limiting memory usage. "+
-		"The limit can't exceed the corresponding -search.maxUniqueTimeseries limit on vmstorage, it can be only set to lower values.")
+		"The limit can't exceed the corresponding value set in vmstorage if the `-search.maxUniqueTimeseries` flag is explicitly defined there.")
```
Comment (hagen1778) [inline app/vmselect/prometheus/prometheus.go:57]:
```suggestion
		"The limit can't exceed the explicitly set corresponding value `-search.maxUniqueTimeseries` on vmstorage side.")
```

**Review by hagen1778 — APPROVED** [review]
LGTM.
Do you think it is worth adding integration tests to verify that query and labels API can override the limits?

**Review by jiekun — APPROVED** [review]

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
Thanks for contribution!

**Pull request merged**